here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/AbstractTask.java

```diff
@@ -54,16 +51,23 @@ public abstract class AbstractTask<RESULT, SELF extends AbstractTask<RESULT, SEL
     implements INakshaBound, UncaughtExceptionHandler {
 
   private static final Logger log = LoggerFactory.getLogger(AbstractTask.class);
-  static int thresholdPerProcessor;
-  static int percentageThresholdPerPrinciple;
+  private static int maxParallelRequestsPerCPU = 50;
+  private static int maxPctParallelRequestsPerPrincipal = 100;
 
   public static void initConcurrencyLimits(int cpuThreshHold, int perPrincipalThreshold) {
-    thresholdPerProcessor = cpuThreshHold;
-    percentageThresholdPerPrinciple = perPrincipalThreshold;
+    maxParallelRequestsPerCPU = cpuThreshHold;
+    maxPctParallelRequestsPerPrincipal = perPrincipalThreshold;
   }
 
-  private static final Map<String, AtomicInteger> authorUsageMap = new ConcurrentHashMap<>();
+  /**
+   * The soft-limit of tasks to run concurrently. This limit does normally not apply to child tasks.
+   */
+  public static final AtomicLong limit =
+      new AtomicLong((long) Runtime.getRuntime().availableProcessors() * maxParallelRequestsPerCPU);
 
+  private static final double principalThreshold = (double) maxPctParallelRequestsPerPrincipal / 100;
+  private static final int principalLimit = (int) (limit.get() * principalThreshold);
+  private static final Map<String, AtomicInteger> principalUsageMap = new ConcurrentHashMap<>();
```

Comment:
Please make this explicitly an ConcurrentHashMap, do not downcast this to a Map, this only hides a important information: It is used concurrently! Apart of this, it later causes you to use a wrong access pattern, because important methods are missing.
Please annotate the entries with @NotNull, as neither the key nor the value must be null.

Reply:
Fixed.